src/track/beats.h

```diff
@@ -54,124 +45,232 @@ class Beats : public QObject {
+    using iterator = BeatIterator;
+
+    static const QString BEAT_MAP_VERSION;
+    static const QString BEAT_GRID_1_VERSION;
+    static const QString BEAT_GRID_2_VERSION;
+    static const QString BEATS_VERSION;
+    Beat findNthBeat(FramePos frame, int offset) const;
+    Beat findNextBeat(FramePos frame) const;
+    Beat findPrevBeat(FramePos frame) const;
+    Bpm getGlobalBpm() const;
+    bool isValid() const;
+    void setSampleRate(int sampleRate);
+    SINT getSampleRate() const {
+        return m_iSampleRate;
+    }
+    void setDurationSeconds(double duration);
+    int numBeatsInRange(FramePos startFrame, FramePos endFrame) const;
+    QByteArray toProtobuf() const;
+    QString getVersion() const;
+    QString getSubVersion() const;
+    void setSubVersion(const QString& subVersion);
+    Bpm calculateBpm(const Beat& startBeat,
+            const Beat& stopBeat) const;
+    void scale(enum BPMScale scale);
+    FramePos findNBeatsFromFrame(FramePos fromFrame, double beats) const;
+    bool findPrevNextBeats(FramePos frame,
+            FramePos* pPrevBeatFrame,
+            FramePos* pNextBeatFrame) const;
+    void setGrid(Bpm dBpm, FramePos firstBeatFrame = kStartFramePos);
+    FramePos findClosestBeat(FramePos frame) const;
+    std::unique_ptr<BeatsInternal::iterator> findBeats(
+            FramePos startFrame, FramePos stopFrame) const;
+    Bpm getBpmAroundPosition(FramePos curFrame, int n) const;
+    void setSignature(TimeSignature sig, int downbeatIndex);
+    void translate(FrameDiff_t numFrames);
+    void setBpm(Bpm bpm, int beatIndex = 0);
+    inline int size() {
+        return m_beats.size();
+    }
+    FramePos getFirstBeatPosition() const;
+    FramePos getLastBeatPosition() const;
+    Beat getBeatAtIndex(int index) {
+        return m_beats.at(index);
+    }
+    void setAsDownbeat(int beatIndex);
+
+  private:
+    void updateBpm();
+    void scaleDouble();
+    void scaleTriple();
+    void scaleQuadruple();
+    void scaleHalve();
+    void scaleThird();
+    void scaleFourth();
+    void scaleMultiple(uint multiple);
+    void scaleFraction(uint fraction);
+    void generateBeatsFromMarkers();
+    void clearMarkers();
+
+    QString m_subVersion;
+    Bpm m_bpm;
+    BeatList m_beats;
+    track::io::Beats m_beatsProto;
+    int m_iSampleRate;
+    double m_dDurationSeconds;
+    friend QDebug operator<<(QDebug dbg, const BeatsInternal& arg);
+};
+
+/// Beats is a class for BPM and beat management classes.
+/// It stores beats information including beats position, down beats position,
+/// phrase beat position and changes in tempo.
+class Beats final : public QObject {
+    Q_OBJECT
+  public:
+    /// Initialize beats with only the track pointer.
+    explicit Beats(const Track* track);
+    /// The source of this byte array is the serialized representation of beats
+    /// generated by the protocol buffer and stored in the database.
+    Beats(const Track* track, const QByteArray& byteArray);
```

Comment:
The visibility of all constructors should be changed from `public`  to `final` with `Track` as a friend class. Only the Track itself is allowed to construct new Beats object by passing `this` as the first parameter.